webapp/loggedin/modules/users/user/actions.js

```diff
@@ -0,0 +1,130 @@
+}
+
+const _validateUser = user => async dispatch => {
+  const validation = await UserValidator.validateUser(user)
+  const userValidated = User.assocValidation(validation)(user)
+  dispatch({ type: userUpdate, user: userValidated })
+  return userValidated
+}
+
+export const updateUserProp = (key, value) => (dispatch, getState) => {
+  const state = getState()
+  const userToUpdate = UserViewState.getUser(state)
+
+  const userUpdated = User.assocProp(key, value)(userToUpdate)
+
+  dispatch(_validateUser(userUpdated))
+}
+
+export const updateUserProfilePicture = profilePicture => async dispatch =>
+  dispatch({ type: userProfilePictureUpdate, profilePicture })
+
+export const saveUser = history => async (dispatch, getState) => {
+  const state = getState()
+  const user = AppState.getUser(state)
+  const userToUpdate = UserViewState.getUser(state)
+  const profilePicture = UserViewState.getProfilePicture(state)
+  const surveyId = SurveyState.getSurveyId(state)
+
+  const userValidated = await dispatch(_validateUser(userToUpdate))
+
+  if (Validation.isObjValid(userValidated)) {
+    const editingSelf = User.isEqual(user)(userToUpdate)
+    try {
+      dispatch(showAppLoader())
+
+      const formData = new FormData()
+      formData.append(User.keys.uuid, User.getUuid(userToUpdate))
+      formData.append(User.keys.name, User.getName(userToUpdate))
+      formData.append(User.keys.email, User.getEmail(userToUpdate))
+      formData.append(User.keys.groupUuid, User.getGroupUuid(userToUpdate))
+      if (profilePicture) {
+        formData.append('file', profilePicture)
+      }
+
+      await axios.put(`/api${editingSelf ? '' : `/survey/${surveyId}`}/user/${User.getUuid(userToUpdate)}`, formData)
+
+      // Update user in redux state if self
+      if (User.isEqual(user)(userToUpdate)) {
+        dispatch(setUser(userToUpdate))
+      }
+
+      dispatch(
+        showNotification('usersView.updateUserConfirmation', {
+          name: User.getName(userToUpdate),
+        }),
+      )
+
+      if (!editingSelf) {
+        history.push(appModuleUri(userModules.users))
+      }
+    } finally {
+      dispatch(hideAppLoader())
+    }
+  }
+}
+
+export const removeUser = history => async (dispatch, getState) => {
+  dispatch(showAppLoader())
+
+  const state = getState()
+  const lang = AppState.getLang(state)
+  const surveyId = SurveyState.getSurveyId(state)
+  const surveyInfo = SurveyState.getSurveyInfo(state)
+  const userToUpdate = UserViewState.getUser(state)
+
+  try {
+    await axios.delete(`/api/survey/${surveyId}/user/${User.getUuid(userToUpdate)}`)
+
+    dispatch(
+      showNotification('userView.removeUserConfirmation', {
+        user: User.getName(userToUpdate),
+        survey: Survey.getLabel(surveyInfo, lang),
+      }),
+    )
+    history.push(appModuleUri(userModules.users))
+  } catch (error) {
+    dispatch(showNotification('appErrors.generic', { text: error }, AppNotificationState.severity.error))
+  }
```

Comment:
isn't this covered by the error middleware?

Reply:
yes, right!